lib/Yield/Assembly/DisassemblerInstance.cpp

```diff
@@ -0,0 +1,338 @@
+/// \file DisassemblerInstance.cpp
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <map>
+#include <string>
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/MC/MCInst.h"
+#include "llvm/MC/MCTargetOptions.h"
+#include "llvm/Support/TargetRegistry.h"
+#include "llvm/Support/TargetSelect.h"
+
+#include "revng/Support/Debug.h"
+#include "revng/Yield/Assembly/Assembly.h"
+#include "revng/Yield/Assembly/DisassemblerInstance.h"
+
+namespace options {
+
+static bool UseIntelSyntax = true;
+
+enum class ImmediateStyles { Decimal, CHexadecimal, AsmHexadecimal };
+static ImmediateStyles ImmediateStyle = ImmediateStyles::AsmHexadecimal;
+
+static bool ShouldSymbolizeOperands = false;
+
+} // namespace options
+
+/// \note: this might cause multithreading problems.
+static void ensureDisassemblersWereInitializedOnce() {
+  static bool WereTheyInitialized = false;
+  if (!WereTheyInitialized) {
+    llvm::InitializeAllTargetInfos();
+    llvm::InitializeAllTargetMCs();
+    llvm::InitializeAllDisassemblers();
+
+    WereTheyInitialized = true;
+  }
+}
+
+using DI = DisassemblerInstance;
+DI::DisassemblerInstance(MetaAddressType::Values AddressType) {
+  ensureDisassemblersWereInitializedOnce();
+
+  auto LLVMArchitecture = MetaAddressType::arch(AddressType);
+  revng_assert(LLVMArchitecture.hasValue(),
+               "Impossible to create a disassembler for a non-code section");
+  auto Architecture = llvm::Triple::getArchTypeName(*LLVMArchitecture);
+
+  // Workaround for ARM
+  if (*LLVMArchitecture == llvm::Triple::ArchType::arm)
+    Architecture = "armv7";
+
+  std::string CPUFeatures = "";
+  if (AddressType == MetaAddressType::Code_arm_thumb)
+    CPUFeatures += "+thumb-mode";
```

Comment:
Can you push this logic close to `MetaAddressType` (`MetaAddressType::getLLVMCPUFeatures(MetaAddressType::Values)`)?